### ISSUE TITLE
Question about t_matrix

### ISSUE THREAD

**Issue by DarrenQu** [yifanlu0227/CoAlign#9]
Is the purpose of t_matrix to rotate the CAVs feature to the ego's space?
Thank you!

![image](https://github.com/yifanlu0227/CoAlign/assets/78702377/3a224303-6315-4f96-923f-27d544c9a83c)

**Comment by yifanlu0227** [comment]
Yes, exactly including translation and rotation.

**Comment by DarrenQu** [comment]
Thank you! I have more questions.
If there is only ego car, is t_matrix a 2x3 matrix of all zeros? 
If there are three cars, say ego and 2CAVs, what's the value of t_matrix? Is it three 2x3 matrices？Transfer the features of the three cars to the ego space respectively?

**Comment by yifanlu0227** [comment]
If there is only ego car, then `N=record_len[b]` would be 1. `t_matrix` retrieve from `pairwise_t_matrix` would have shape [1,1,2,3]. Particularly, `t_matrix[0,0] = array([[1,0,0],[0,1,0]])`, the identity affine tranformation matrix.

If there are three cars, then `N=record_len[b]` would be 3. `t_matrix` retrieve from `pairwise_t_matrix` would have shape [3,3,2,3]. `t_matrix[i,i]` is still the identity affine transformation matrix while `t_matrix[i,j]` is not. 

You can first investigate how original `pairwise_t_matrix` is generated. see [here](https://github.com/yifanlu0227/CoAlign/blob/6f16c06b266d0af288bf8546048b35cfb50fdb3a/opencood/utils/transformation_utils.py#L22). It has output shape [L,L,4,4], where L is the max_cav number. Then when collating samples in a batch, the `pairwise_t_matrix` would have shape [B, L, L, 4, 4], where B is the batchsize. 

Note that the pairwise_t_matrix (with shape [B, L, L, 4, 4]) has accurate physical meanings, it is pairwise tranformation matrix for different coodinate. But when used for image(feature) affine transformation, it should be normalized and abandon z-axis component. See [here](https://github.com/yifanlu0227/CoAlign/blob/6f16c06b266d0af288bf8546048b35cfb50fdb3a/opencood/models/point_pillar_baseline.py#L115C16-L115C16). The output of `normalize_pairwise_tfm()` would have shape [B, L, L, 2, 3], which turned out to be the input `pairwise_t_matrix` in your image. There is some confusion in naming. It is more appropriate to change `pairwise_t_matrix` to `normalized_pairwise_t_matrix` in your image.

**Comment by DarrenQu** [comment]
That makes sense, many thanks for addressing my questions.

**Comment by dwc13** [comment]
I have a question about the ```normalize_pairwise_tfm()``` function, that you have mentioned above to [here](https://github.com/yifanlu0227/CoAlign/blob/6f16c06b266d0af288bf8546048b35cfb50fdb3a/opencood/models/point_pillar_baseline.py#L115).

Within this [function](https://github.com/yifanlu0227/CoAlign/blob/6617aa58752df7adfe14792fc20701649d0e307a/opencood/utils/transformation_utils.py#L69), the description talks about normalizing the values. However, you are multiplying by 2 which confuses me in the logical reasoning in this normalization process. The idea of normalization is to scale it, and in this case point clouds to feature map sizes, which is done by discrete_ratio, downsample_rate, height, and width.

Thus, what is the mathematical reason that you are multiplying by 2?

**Comment by yifanlu0227** [comment]
@dwc13 Hi, when we talk about the normalize affine matrix, we should always consider the `torch.nn.functional.affine_grid` and `torch.nn.functional.grid_sample` functions, which accept the normalized affine matrix as input to perform the spatial warping.

`torch.nn.functional.affine_grid` will generate a sampling grid $G = M \cdot G_{identity}$, where $M$ is your normalized affine matrix and  $G_{identity}$ is the identity sampling grid. $\cdot$ is homogeneous matrix multiplication.

`torch.nn.functional.grid_sample` will further sample pixel values according to the give grid. $f({\rm sampling\ grid} \rightarrow {\rm pixel\ coordinate}): [-1, 1] \times [-1, 1] \rightarrow [-H/2,H/2] \times [-W/2, W/2]$. 


For example, $G_{identity}$ is 

```
>>> theta = torch.Tensor([[[1,0,0],[0,1,0]]])
>>> size = torch.Size([1,1,4,4]) # suppose feature map shape is 4 by 4
>>> G_identity = F.affine_grid(theta,size)
>>> G_identity
tensor([[[[-0.7500, -0.7500],
          [-0.2500, -0.7500],
          [ 0.2500, -0.7500],
          [ 0.7500, -0.7500]],

         [[-0.7500, -0.2500],
          [-0.2500, -0.2500],
          [ 0.2500, -0.2500],
          [ 0.7500, -0.2500]],

         [[-0.7500,  0.2500],
          [-0.2500,  0.2500],
          [ 0.2500,  0.2500],
          [ 0.7500,  0.2500]],

         [[-0.7500,  0.7500],
          [-0.2500,  0.7500],
          [ 0.2500,  0.7500],
          [ 0.7500,  0.7500]]]])
>>> feature_map = torch.full((1,1,4,4), 10).float()
>>> F.grid_sample(feature_map, G_identity) # no warping is performed
tensor([[[[10., 10., 10., 10.],
          [10., 10., 10., 10.],
          [10., 10., 10., 10.],
          [10., 10., 10., 10.]]]])
``` 

Now we consider the one extreme situation, where the distance between two agents in x-axis corresponds to feature map' physical dimension. In this case, the warped feature map in ego's view should be all empty.

Now in [normalization step](https://github.com/yifanlu0227/CoAlign/blob/6617aa58752df7adfe14792fc20701649d0e307a/opencood/utils/transformation_utils.py#L69), we have `pairwise_t_matrix[...,0,2] == (downsample_rate * discrete_ratio * W)`, thus

`pairwise_t_matrix[...,0,2] = pairwise_t_matrix[...,0,2] / (downsample_rate * discrete_ratio * W) * 2 = 2`, 

But we do not multiply 2 on the translation component, the normalized affine matrix will be:
```
>>> M = torch.Tensor([[[1,0,1],[0,1,0]]]) # note here the relative translation component is 1.
>>> size = torch.Size([1,1,4,4])
>>> G = F.affine_grid(M,size)
>>> G
tensor([[[[ 0.2500, -0.7500],
          [ 0.7500, -0.7500],
          [ 1.2500, -0.7500],
          [ 1.7500, -0.7500]],

         [[ 0.2500, -0.2500],
          [ 0.7500, -0.2500],
          [ 1.2500, -0.2500],
          [ 1.7500, -0.2500]],

         [[ 0.2500,  0.2500],
          [ 0.7500,  0.2500],
          [ 1.2500,  0.2500],
          [ 1.7500,  0.2500]],

         [[ 0.2500,  0.7500],
          [ 0.7500,  0.7500],
          [ 1.2500,  0.7500],
          [ 1.7500,  0.7500]]]])

>>> F.grid_sample(feature_map, G) 
tensor([[[[10., 10.,  0.,  0.],
          [10., 10.,  0.,  0.],
          [10., 10.,  0.,  0.],
          [10., 10.,  0.,  0.]]]])
```

You can see only half of the feature map is translated. This is inaccurate. There is still half of the grids have value in $[-1,1] \times [-1, 1]$.


Now we multiply the translation component with 2, all the behaviour becomes correct.
```
>>> M = torch.Tensor([[[1,0,2],[0,1,0]]]) # note here the relative translation component is 1.
>>> G = F.affine_grid(M,size)
>>> G
tensor([[[[ 1.2500, -0.7500],
          [ 1.7500, -0.7500],
          [ 2.2500, -0.7500],
          [ 2.7500, -0.7500]],

         [[ 1.2500, -0.2500],
          [ 1.7500, -0.2500],
          [ 2.2500, -0.2500],
          [ 2.7500, -0.2500]],

         [[ 1.2500,  0.2500],
          [ 1.7500,  0.2500],
          [ 2.2500,  0.2500],
          [ 2.7500,  0.2500]],

         [[ 1.2500,  0.7500],
          [ 1.7500,  0.7500],
          [ 2.2500,  0.7500],
          [ 2.7500,  0.7500]]]])

>>> F.grid_sample(feature_map, G)
tensor([[[[0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.],
          [0., 0., 0., 0.]]]])
```
This is correct.

I'm trying to explain as clearly as I can, but I think you need to read these two functions by yourself to get a full understanding of the spatial warping procedure :)
